include/boost/url/impl/any_allocator.hpp

```diff
@@ -0,0 +1,227 @@
+  template <typename A, class Enable = void> struct allocator_is_always_equal : std::is_empty<A> {};
+
+  template <typename A>
+  struct allocator_is_always_equal<A, boost::enable_if_t<has_is_always_equal<A>::value>> : A::is_always_equal {};
+
+  template <typename A>
+  struct allocator_is_always_equal<A, boost::enable_if_t<has_is_always_equal<std::allocator_traits<A>>::value && !std::allocator_traits<A>::value>> : std::allocator_traits<A>::is_always_equal {};
+
+}
+
+
+// base class to store the underlying allocator bound to `base`
+template <class T>
+struct any_allocator<T>::base
+{
+  std::atomic<std::size_t> refs{1};
+
+  virtual
+  ~base()
+  = default;
+
+  explicit
+  base(std::size_t r) : refs(r) {
+
+  }
+
+  virtual
+  char*
+  allocate(std::size_t n) = 0;
+
+  virtual
+  void
+  deallocate(char*p, std::size_t n) = 0;
+
+  template <class U>
+  void
+  deallocate(U* p, std::size_t n) {
+      deallocate(reinterpret_cast<char*>(p), n * sizeof(T));
+  }
+
+  void
+  deallocate(char*p) {
+      deallocate(p, 1);
+  }
+
+  template <class U>
+  void
+  deallocate(U* p) {
+      deallocate(p, 1);
+  }
+
+  virtual void destroy_self() noexcept = 0;
+
+};
+
+template <class T>
+any_allocator<T>::
+~any_allocator()
+{
+  if( p_ && --p_->refs == 0)
+    p_->destroy_self();
+}
+
+template <class T>
+any_allocator<T>&
+any_allocator<T>::
+operator=(
+    any_allocator<T> const& other) noexcept
+{
+  if (this == &other)
+    return *this;
+  if(other.p_)
+    ++other.p_->refs;
+  if( p_ && --p_->refs == 0)
+    p_->destroy_self();
+  p_ = other.p_;
+  return *this;
+}
+
+template <class T>
+template <class Allocator>
+any_allocator<T>::
+any_allocator(Allocator const& a) noexcept
+{
+  class impl;
+
+  // an allocator we can use to allocator `base`
+  using char_allocator_type = typename
+      detail::allocator_traits<
+          Allocator>::template
+      rebind_alloc<char>;
+
+  class impl : public base
+  {
+    char_allocator_type a_;
+
+  public:
+    ~impl() {}
+
+    explicit
+    impl(char_allocator_type const& a)
+        : base(1), a_(a)
+    {
+    }
+
+    char*
+    allocate(std::size_t n) override
+    {
+      /*
+         if( align <= alignof(char) ) // rebind to char
+         else if( align <= alignof(short) ) // rebind to short
+         else if ...
+         else if( align <= alignof(max_align_t) ) // rebind to max_align_t
+         else throw ...
+      */
+      return a_.allocate(n);
+    }
+
+    void
+    deallocate(char* p, std::size_t n) {
+      a_.deallocate(p, n);
+    }
+
+    void
+    destroy_self() noexcept override
+    {
+      base::deallocate(this);
+    }
+
+    bool
+    is_always_equal() const {
+      return detail::allocator_is_always_equal<char_allocator_type>::value;
+    }
+  };
+
+  char_allocator_type al(a);
+  p_ = ::new(al.allocate(1)) impl(al);
+}
+
+template <class T>
+any_allocator<T>::
+any_allocator(
+    any_allocator const& other) noexcept : p_(other.p_) {
```

Comment:
can we please put opening braces on their own line? otherwise it is harder to debug